django/db/models/fields/generated.py

```diff
@@ -17,6 +19,8 @@ class GeneratedField(Field):
     output_field = None
 
     def __init__(self, *, expression, db_persist=None, output_field=None, **kwargs):
+        if output_field is None:
+            raise FieldError("Expression doesn't contain an explicit type. You must set output_field.")
```

Comment:
I doesn't need to be explicit set, it can be inferred, see `contribute_to_class`.

Reply:
I see. So when does it need to be explicitly set? do I need to check if the expression contains `Concat`, after ```self.output_field = (
            self._output_field
            if self._output_field is not None
            else self._resolved_expression.output_field
        )```
...in `contribute_to_class`?

Reply:
It needs to be explicit set when the resolved field doesn't pass the check added in `_check_output_field`. The returned check errors should have a `hint` that an explicit `output_field` should be set to address the check failure.

For example, in the reported case (by the way this should be assigned to ticket-34944 and not ticket-34954) when a `CharField(max_length=None)` is inferred the resulting check error should have a hint pointing at setting an explicit `output_field`.

Reply:
Okay, let me finish the `_check_output_field` first. Sorry for typo